File: game/scene/Scene.lua
-- [ts]: Scene.ts
local ____lualib = require("lualib_bundle") -- 1
local __TS__Class = ____lualib.__TS__Class -- 1
local __TS__New = ____lualib.__TS__New -- 1
local ____exports = {} -- 1
local circleVertsAt, drawRing -- 1
local ____Dora = require("Dora") -- 2
local Color = ____Dora.Color -- 2
local DrawNode = ____Dora.DrawNode -- 2
local Vec2 = ____Dora.Vec2 -- 2
local ____RNG = require("game.utils.RNG") -- 3
local RNG = ____RNG.RNG -- 3
local ____MathUtils = require("game.utils.MathUtils") -- 4
local clamp = ____MathUtils.clamp -- 4
function circleVertsAt(center, radius, segments) -- 156
	local verts = {} -- 157
	do -- 157
		local i = 0 -- 158
		while i < segments do -- 158
			local a = i / segments * math.pi * 2 -- 159
			verts[#verts + 1] = Vec2( -- 160
				center.x + math.cos(a) * radius, -- 160
				center.y + math.sin(a) * radius -- 160
			) -- 160
			i = i + 1 -- 158
		end -- 158
	end -- 158
	return verts -- 162
end -- 162
function drawRing(draw, center, radius, segments, width, color) -- 165
	do -- 165
		local i = 0 -- 166
		while i < segments do -- 166
			local a1 = i / segments * math.pi * 2 -- 167
			local a2 = (i + 1) / segments * math.pi * 2 -- 168
			draw:drawSegment( -- 169
				Vec2( -- 169
					center.x + math.cos(a1) * radius, -- 169
					center.y + math.sin(a1) * radius -- 169
				), -- 169
				Vec2( -- 169
					center.x + math.cos(a2) * radius, -- 169
					center.y + math.sin(a2) * radius -- 169
				), -- 169
				width, -- 169
				color -- 169
			) -- 169
			i = i + 1 -- 166
		end -- 166
	end -- 166
end -- 166
local WORLD_SIZE = 2000 -- 6
local HALF = WORLD_SIZE / 2 -- 7
local TILE_STEP = 160 -- 8
local DECOR_COUNT = 110 -- 9
local DECOR_SEED = 20240520 -- 10
____exports.Scene = __TS__Class() -- 12
local Scene = ____exports.Scene -- 12
Scene.name = "Scene" -- 12
function Scene.prototype.____constructor(self, root) -- 18
	self.spawnRng = __TS__New(RNG, 2882400001) -- 64
	self.root = root -- 19
	self.background = DrawNode() -- 20
	self.decorDraw = DrawNode() -- 21
	self.boundsDraw = DrawNode() -- 22
	self:drawBackground() -- 23
	self:drawBounds() -- 24
	self:buildDecor() -- 25
	self.background:addTo(self.root) -- 26
	self.decorDraw:addTo(self.root) -- 27
	self.boundsDraw:addTo(self.root) -- 28
end -- 18
function Scene.prototype.getWorldBounds(self) -- 31
	return {minX = -HALF, maxX = HALF, minY = -HALF, maxY = HALF} -- 32
end -- 31
function Scene.prototype.getSpawnPointNearEdge(self, worldCenter, margin) -- 36
	do -- 36
		local attempt = 0 -- 37
		while attempt < 8 do -- 37
			local a = self.spawnRng:range(0, math.pi * 2) -- 38
			local r = self.spawnRng:range(HALF - 160, HALF - 50) -- 39
			local x = math.cos(a) * r -- 40
			local y = math.sin(a) * r -- 41
			local dx = x - worldCenter.x -- 42
			local dy = y - worldCenter.y -- 43
			if dx * dx + dy * dy >= margin * margin then -- 43
				return Vec2(x, y) -- 45
			end -- 45
			attempt = attempt + 1 -- 37
		end -- 37
	end -- 37
	local a2 = self.spawnRng:range(0, math.pi * 2) -- 49
	local x2 = clamp( -- 50
		worldCenter.x + math.cos(a2) * margin, -- 50
		-HALF + 40, -- 50
		HALF - 40 -- 50
	) -- 50
	local y2 = clamp( -- 51
		worldCenter.y + math.sin(a2) * margin, -- 51
		-HALF + 40, -- 51
		HALF - 40 -- 51
	) -- 51
	return Vec2(x2, y2) -- 52
end -- 36
function Scene.prototype.isInside(self, pos) -- 55
	return math.abs(pos.x) <= HALF and math.abs(pos.y) <= HALF -- 56
end -- 55
function Scene.prototype.update(self, dt) -- 59
end -- 59
function Scene.prototype.drawBackground(self) -- 66
	self.background:clear() -- 67
	local c = Color(13, 24, 26, 255) -- 68
	self.background:drawPolygon( -- 69
		{ -- 70
			Vec2(-HALF, -HALF), -- 70
			Vec2(HALF, -HALF), -- 70
			Vec2(HALF, HALF), -- 70
			Vec2(-HALF, HALF) -- 70
		}, -- 70
		c -- 71
	) -- 71
	do -- 71
		local gx = -HALF -- 74
		while gx < HALF do -- 74
			do -- 74
				local gy = -HALF -- 75
				while gy < HALF do -- 75
					local odd = (math.floor((gx + HALF) / TILE_STEP) + math.floor((gy + HALF) / TILE_STEP)) % 2 == 0 -- 76
					local pad = 4 -- 77
					self.background:drawPolygon( -- 78
						{ -- 79
							Vec2(gx + pad, gy + pad), -- 79
							Vec2(gx + TILE_STEP - pad, gy + pad), -- 79
							Vec2(gx + TILE_STEP - pad, gy + TILE_STEP - pad), -- 79
							Vec2(gx + pad, gy + TILE_STEP - pad) -- 79
						}, -- 79
						Color(odd and 17 or 15, odd and 31 or 28, odd and 32 or 30, 255) -- 80
					) -- 80
					gy = gy + TILE_STEP -- 75
				end -- 75
			end -- 75
			gx = gx + TILE_STEP -- 74
		end -- 74
	end -- 74
	self.background:drawDot( -- 85
		Vec2.zero, -- 85
		190, -- 85
		Color(23, 52, 48, 255) -- 85
	) -- 85
	drawRing( -- 86
		self.background, -- 86
		Vec2.zero, -- 86
		190, -- 86
		32, -- 86
		4, -- 86
		Color(61, 104, 91, 180) -- 86
	) -- 86
	drawRing( -- 87
		self.background, -- 87
		Vec2.zero, -- 87
		116, -- 87
		24, -- 87
		2, -- 87
		Color(203, 156, 67, 150) -- 87
	) -- 87
	do -- 87
		local i = 0 -- 88
		while i < 8 do -- 88
			local a = i / 8 * math.pi * 2 -- 89
			local p1 = Vec2( -- 90
				math.cos(a) * 126, -- 90
				math.sin(a) * 126 -- 90
			) -- 90
			local p2 = Vec2( -- 91
				math.cos(a) * 176, -- 91
				math.sin(a) * 176 -- 91
			) -- 91
			self.background:drawSegment( -- 92
				p1, -- 92
				p2, -- 92
				5, -- 92
				Color(172, 127, 53, 120) -- 92
			) -- 92
			i = i + 1 -- 88
		end -- 88
	end -- 88
end -- 66
function Scene.prototype.drawBounds(self) -- 96
	self.boundsDraw:clear() -- 97
	local shadow = Color(0, 0, 0, 180) -- 98
	local c = Color(197, 145, 56, 255) -- 99
	self.boundsDraw:drawSegment( -- 100
		Vec2(-HALF, -HALF), -- 100
		Vec2(HALF, -HALF), -- 100
		12, -- 100
		shadow -- 100
	) -- 100
	self.boundsDraw:drawSegment( -- 101
		Vec2(HALF, -HALF), -- 101
		Vec2(HALF, HALF), -- 101
		12, -- 101
		shadow -- 101
	) -- 101
	self.boundsDraw:drawSegment( -- 102
		Vec2(HALF, HALF), -- 102
		Vec2(-HALF, HALF), -- 102
		12, -- 102
		shadow -- 102
	) -- 102
	self.boundsDraw:drawSegment( -- 103
		Vec2(-HALF, HALF), -- 103
		Vec2(-HALF, -HALF), -- 103
		12, -- 103
		shadow -- 103
	) -- 103
	self.boundsDraw:drawSegment( -- 104
		Vec2(-HALF, -HALF), -- 104
		Vec2(HALF, -HALF), -- 104
		4, -- 104
		c -- 104
	) -- 104
	self.boundsDraw:drawSegment( -- 105
		Vec2(HALF, -HALF), -- 105
		Vec2(HALF, HALF), -- 105
		4, -- 105
		c -- 105
	) -- 105
	self.boundsDraw:drawSegment( -- 106
		Vec2(HALF, HALF), -- 106
		Vec2(-HALF, HALF), -- 106
		4, -- 106
		c -- 106
	) -- 106
	self.boundsDraw:drawSegment( -- 107
		Vec2(-HALF, HALF), -- 107
		Vec2(-HALF, -HALF), -- 107
		4, -- 107
		c -- 107
	) -- 107
end -- 96
function Scene.prototype.buildDecor(self) -- 111
	self.decorDraw:clear() -- 112
	local rng = __TS__New(RNG, DECOR_SEED) -- 113
	do -- 113
		local i = 0 -- 114
		while i < DECOR_COUNT do -- 114
			local x = rng:range(-HALF + 30, HALF - 30) -- 115
			local y = rng:range(-HALF + 30, HALF - 30) -- 116
			if rng:chance(0.45) then -- 116
				local r = rng:range(2, 5) -- 119
				self.decorDraw:drawDot( -- 120
					Vec2(x, y), -- 120
					r + 4, -- 120
					Color(45, 136, 105, 24) -- 120
				) -- 120
				self.decorDraw:drawDot( -- 121
					Vec2(x, y), -- 121
					r, -- 121
					Color( -- 121
						64, -- 121
						rng:int(120, 185), -- 121
						122, -- 121
						185 -- 121
					) -- 121
				) -- 121
			elseif rng:chance(0.62) then -- 121
				local a = rng:range(0, math.pi * 2) -- 124
				local len = rng:range(12, 30) -- 125
				local mid = Vec2( -- 126
					x + math.cos(a) * len * 0.5, -- 126
					y + math.sin(a) * len * 0.5 -- 126
				) -- 126
				self.decorDraw:drawSegment( -- 127
					Vec2(x, y), -- 127
					mid, -- 127
					1.5, -- 127
					Color(58, 73, 70, 170) -- 127
				) -- 127
				self.decorDraw:drawSegment( -- 128
					mid, -- 128
					Vec2( -- 128
						mid.x + math.cos(a + 0.7) * len * 0.45, -- 128
						mid.y + math.sin(a + 0.7) * len * 0.45 -- 128
					), -- 128
					1.2, -- 128
					Color(58, 73, 70, 130) -- 128
				) -- 128
			else -- 128
				local r = rng:range(5, 11) -- 131
				self.decorDraw:drawPolygon( -- 132
					{ -- 132
						Vec2(x - r, y - r * 0.4), -- 132
						Vec2(x + r * 0.7, y - r), -- 132
						Vec2(x + r, y + r * 0.5), -- 132
						Vec2(x - r * 0.5, y + r) -- 132
					}, -- 132
					Color(53, 65, 65, 220) -- 132
				) -- 132
			end -- 132
			i = i + 1 -- 114
		end -- 114
	end -- 114
	local sites = { -- 136
		Vec2(-560, 420), -- 136
		Vec2(620, 500), -- 136
		Vec2(-610, -520), -- 136
		Vec2(560, -470) -- 136
	} -- 136
	for ____, p in ipairs(sites) do -- 137
		self:drawRuin(p) -- 137
	end -- 137
end -- 111
function Scene.prototype.drawRuin(self, p) -- 140
	drawRing( -- 141
		self.decorDraw, -- 141
		p, -- 141
		72, -- 141
		16, -- 141
		5, -- 141
		Color(79, 91, 83, 210) -- 141
	) -- 141
	drawRing( -- 142
		self.decorDraw, -- 142
		p, -- 142
		48, -- 142
		12, -- 142
		2, -- 142
		Color(172, 127, 53, 140) -- 142
	) -- 142
	do -- 142
		local i = 0 -- 143
		while i < 4 do -- 143
			local a = i * math.pi / 2 + 0.25 -- 144
			local base = Vec2( -- 145
				p.x + math.cos(a) * 78, -- 145
				p.y + math.sin(a) * 78 -- 145
			) -- 145
			self.decorDraw:drawSegment( -- 146
				base, -- 146
				Vec2(base.x, base.y + 42), -- 146
				9, -- 146
				Color(64, 75, 71, 255) -- 146
			) -- 146
			self.decorDraw:drawDot( -- 147
				Vec2(base.x, base.y + 43), -- 147
				8, -- 147
				Color(89, 99, 90, 255) -- 147
			) -- 147
			i = i + 1 -- 143
		end -- 143
	end -- 143
end -- 140
local function circleVerts(radius, segments) -- 152
	return circleVertsAt(Vec2.zero, radius, segments) -- 153
end -- 152
return ____exports -- 152

File: game/scene/Scene.ts
// 场景：世界背景、网格、边界、装饰与刷怪点
import { Color, DrawNode, Node, Vec2 } from 'Dora';
import { RNG } from 'game/utils/RNG';
import { clamp } from 'game/utils/MathUtils';

const WORLD_SIZE = 2000;
const HALF = WORLD_SIZE / 2;
const TILE_STEP = 160;
const DECOR_COUNT = 110;
const DECOR_SEED = 20240520;

export class Scene {
	private root: Node.Type;
	private background: DrawNode.Type;
	private decorDraw: DrawNode.Type;
	private boundsDraw: DrawNode.Type;

	constructor(root: Node.Type) {
		this.root = root;
		this.background = DrawNode();
		this.decorDraw = DrawNode();
		this.boundsDraw = DrawNode();
		this.drawBackground();
		this.drawBounds();
		this.buildDecor();
		this.background.addTo(this.root);
		this.decorDraw.addTo(this.root);
		this.boundsDraw.addTo(this.root);
	}

	getWorldBounds(): { minX: number; maxX: number; minY: number; maxY: number } {
		return { minX: -HALF, maxX: HALF, minY: -HALF, maxY: HALF };
	}

	// 从世界边缘带随机生成刷怪点（默认避开玩家视野 margin）
	getSpawnPointNearEdge(worldCenter: Vec2.Type, margin: number): Vec2.Type {
		for (let attempt = 0; attempt < 8; attempt++) {
			const a = this.spawnRng.range(0, Math.PI * 2);
			const r = this.spawnRng.range(HALF - 160, HALF - 50);
			const x = Math.cos(a) * r;
			const y = Math.sin(a) * r;
			const dx = x - worldCenter.x;
			const dy = y - worldCenter.y;
			if (dx * dx + dy * dy >= margin * margin) {
				return Vec2(x, y);
			}
		}
		// 兜底：玩家向外 margin 方向（clamp 在世界内）
		const a2 = this.spawnRng.range(0, Math.PI * 2);
		const x2 = clamp(worldCenter.x + Math.cos(a2) * margin, -HALF + 40, HALF - 40);
		const y2 = clamp(worldCenter.y + Math.sin(a2) * margin, -HALF + 40, HALF - 40);
		return Vec2(x2, y2);
	}

	isInside(pos: Vec2.Type): boolean {
		return Math.abs(pos.x) <= HALF && Math.abs(pos.y) <= HALF;
	}

	update(dt: number): void {
		// 目前无动态场景元素
	}

	// ---------- 内部 ----------
	private spawnRng = new RNG(0xabcdef01);

	private drawBackground(): void {
		this.background.clear();
		const c = Color(13, 24, 26, 255);
		this.background.drawPolygon(
			[Vec2(-HALF, -HALF), Vec2(HALF, -HALF), Vec2(HALF, HALF), Vec2(-HALF, HALF)],
			c,
		);
		// 不规则石板分区，避免开发调试感的全屏工程网格。
		for (let gx = -HALF; gx < HALF; gx += TILE_STEP) {
			for (let gy = -HALF; gy < HALF; gy += TILE_STEP) {
				const odd = (Math.floor((gx + HALF) / TILE_STEP) + Math.floor((gy + HALF) / TILE_STEP)) % 2 === 0;
				const pad = 4;
				this.background.drawPolygon(
					[Vec2(gx + pad, gy + pad), Vec2(gx + TILE_STEP - pad, gy + pad), Vec2(gx + TILE_STEP - pad, gy + TILE_STEP - pad), Vec2(gx + pad, gy + TILE_STEP - pad)],
					Color(odd ? 17 : 15, odd ? 31 : 28, odd ? 32 : 30, 255),
				);
			}
		}
		// 出生祭坛：让地图中心成为明确的视觉锚点。
		this.background.drawDot(Vec2.zero, 190, Color(23, 52, 48, 255));
		drawRing(this.background, Vec2.zero, 190, 32, 4, Color(61, 104, 91, 180));
		drawRing(this.background, Vec2.zero, 116, 24, 2, Color(203, 156, 67, 150));
		for (let i = 0; i < 8; i++) {
			const a = (i / 8) * Math.PI * 2;
			const p1 = Vec2(Math.cos(a) * 126, Math.sin(a) * 126);
			const p2 = Vec2(Math.cos(a) * 176, Math.sin(a) * 176);
			this.background.drawSegment(p1, p2, 5, Color(172, 127, 53, 120));
		}
	}

	private drawBounds(): void {
		this.boundsDraw.clear();
		const shadow = Color(0, 0, 0, 180);
		const c = Color(197, 145, 56, 255);
		this.boundsDraw.drawSegment(Vec2(-HALF, -HALF), Vec2(HALF, -HALF), 12, shadow);
		this.boundsDraw.drawSegment(Vec2(HALF, -HALF), Vec2(HALF, HALF), 12, shadow);
		this.boundsDraw.drawSegment(Vec2(HALF, HALF), Vec2(-HALF, HALF), 12, shadow);
		this.boundsDraw.drawSegment(Vec2(-HALF, HALF), Vec2(-HALF, -HALF), 12, shadow);
		this.boundsDraw.drawSegment(Vec2(-HALF, -HALF), Vec2(HALF, -HALF), 4, c);
		this.boundsDraw.drawSegment(Vec2(HALF, -HALF), Vec2(HALF, HALF), 4, c);
		this.boundsDraw.drawSegment(Vec2(HALF, HALF), Vec2(-HALF, HALF), 4, c);
		this.boundsDraw.drawSegment(Vec2(-HALF, HALF), Vec2(-HALF, -HALF), 4, c);
	}

	// 固定种子散布装饰（草/石），数量受控
	private buildDecor(): void {
		this.decorDraw.clear();
		const rng = new RNG(DECOR_SEED);
		for (let i = 0; i < DECOR_COUNT; i++) {
			const x = rng.range(-HALF + 30, HALF - 30);
			const y = rng.range(-HALF + 30, HALF - 30);
			if (rng.chance(0.45)) {
				// 发光菌簇
				const r = rng.range(2, 5);
				this.decorDraw.drawDot(Vec2(x, y), r + 4, Color(45, 136, 105, 24));
				this.decorDraw.drawDot(Vec2(x, y), r, Color(64, rng.int(120, 185), 122, 185));
			} else if (rng.chance(0.62)) {
				// 裂纹，由两段不共线笔画构成。
				const a = rng.range(0, Math.PI * 2);
				const len = rng.range(12, 30);
				const mid = Vec2(x + Math.cos(a) * len * 0.5, y + Math.sin(a) * len * 0.5);
				this.decorDraw.drawSegment(Vec2(x, y), mid, 1.5, Color(58, 73, 70, 170));
				this.decorDraw.drawSegment(mid, Vec2(mid.x + Math.cos(a + 0.7) * len * 0.45, mid.y + Math.sin(a + 0.7) * len * 0.45), 1.2, Color(58, 73, 70, 130));
			} else {
				// 破碎石块
				const r = rng.range(5, 11);
				this.decorDraw.drawPolygon([Vec2(x - r, y - r * 0.4), Vec2(x + r * 0.7, y - r), Vec2(x + r, y + r * 0.5), Vec2(x - r * 0.5, y + r)], Color(53, 65, 65, 220));
			}
		}
		// 四处固定遗迹，使移动时景观不再同质。
		const sites = [Vec2(-560, 420), Vec2(620, 500), Vec2(-610, -520), Vec2(560, -470)];
		for (const p of sites) this.drawRuin(p);
	}

	private drawRuin(p: Vec2.Type): void {
		drawRing(this.decorDraw, p, 72, 16, 5, Color(79, 91, 83, 210));
		drawRing(this.decorDraw, p, 48, 12, 2, Color(172, 127, 53, 140));
		for (let i = 0; i < 4; i++) {
			const a = i * Math.PI / 2 + 0.25;
			const base = Vec2(p.x + Math.cos(a) * 78, p.y + Math.sin(a) * 78);
			this.decorDraw.drawSegment(base, Vec2(base.x, base.y + 42), 9, Color(64, 75, 71, 255));
			this.decorDraw.drawDot(Vec2(base.x, base.y + 43), 8, Color(89, 99, 90, 255));
		}
	}
}

function circleVerts(radius: number, segments: number): Vec2.Type[] {
	return circleVertsAt(Vec2.zero, radius, segments);
}

function circleVertsAt(center: Vec2.Type, radius: number, segments: number): Vec2.Type[] {
	const verts: Vec2.Type[] = [];
	for (let i = 0; i < segments; i++) {
		const a = (i / segments) * Math.PI * 2;
		verts.push(Vec2(center.x + Math.cos(a) * radius, center.y + Math.sin(a) * radius));
	}
	return verts;
}

function drawRing(draw: DrawNode.Type, center: Vec2.Type, radius: number, segments: number, width: number, color: Color.Type): void {
	for (let i = 0; i < segments; i++) {
		const a1 = (i / segments) * Math.PI * 2;
		const a2 = ((i + 1) / segments) * Math.PI * 2;
		draw.drawSegment(Vec2(center.x + Math.cos(a1) * radius, center.y + Math.sin(a1) * radius), Vec2(center.x + Math.cos(a2) * radius, center.y + Math.sin(a2) * radius), width, color);
	}
}
